### PR TITLE
Added validation for basic answer

### PR DESCRIPTION
**Pull request by pricea1**
Added validation messages for answers

### What is the context of this PR?
Validation messages have been added to empty basic answer labels

### How to review 
1. Create new page
2. Create answer - no validation message should display
3. Unfocus from answer - validation message should display
4. Create Text & Number answers
5. Add label and unfocus from answer - no validation message should display
6. Edit existing page with existing empty answer label - validation message should display

Works as per prototype - [https://author.netlify.com](https://author.netlify.com)

### CODE REVIEW

File: eq-author-api/schema/tests/validation.test.js [hunk 1]
```diff
@@ -64,16 +64,14 @@ describe("validation", () => {
...
-          errors: [
-            {
-              field: "field",
-              errorCode: "ERR_CODE",
-            },
-          ],
+      const validationErrorInfo = [
+        {
+          id: page.id,
+          type: "page",
```
Comment (tgandrews) [inline eq-author-api/schema/tests/validation.test.js]: Should this type be `pages` as in questionPage? Could this be a constant?

File: eq-author-api/src/validation/schemas/entities.json [hunk 2]
```diff
@@ -1,31 +1,47 @@
...
         }
       },
-      "required": ["id", "title", "pages"]
+      "required": ["id", "title"]
+    },
+    "answer": {
+      "type": "object",
+      "properties": {
+        "id": { "type": "string" },
+        "label": {
```
Comment (tgandrews) [inline eq-author-api/src/validation/schemas/entities.json]: I don't think label has to be populated for all answer types which this is currently matching.

File: eq-author/src/components/Forms/WrappingInput/index.js [hunk 3]
```diff
@@ -48,14 +50,19 @@ class WrappingInput extends React.Component {
...
       <StyleContext
         bold={bold}
         onChange={this.handleChange}
         onKeyDown={this.handleKeyDown}
       >
-        <TextArea {...otherProps} placeholder={placeholder} />
+        <TextArea
+          {...otherProps}
+          placeholder={placeholder}
+          invalid={errorValidationMsg ? "true" : null}
```
Comment (tgandrews) [inline eq-author/src/components/Forms/WrappingInput/index.js]: Can we use booleans here?

File: eq-author/src/App/page/Design/answers/MultipleChoiceAnswer/__snapshots__/MultipleChoiceAnswer.test.js.snap [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
 exports[`MultipleChoiceAnswer exclusive options should render exclusive 1`] = `
-<Connect(withEntityEditor(StatelessBasicAnswer))
+<_class
```
Comment (tgandrews) [inline eq-author/src/App/page/Design/answers/MultipleChoiceAnswer/__snapshots__/MultipleChoiceAnswer.test.js.snap]: I presume this is unintentional

File: eq-author/src/components/Forms/WrappingInput/index.js (hunk 3 above)
Comment (pricea1) [reply eq-author/src/components/Forms/WrappingInput/index.js]: I tried that but got a warning
``
index.js:1384 Warning: Received 'false' for a non-boolean attribute 'invalid'.

If you want to write it to the DOM, pass a string instead: invalid="false" or invalid={value.toString()}.

If you used to conditionally omit it with invalid={condition && value}, pass invalid={condition ? value : undefined} instead.
``

I'll change it so it just takes the errorValidationMsg string.

File: eq-author-api/schema/tests/validation.test.js (hunk 1 above)
Comment (pricea1) [reply eq-author-api/schema/tests/validation.test.js]: Yup, now changed to pages and added as constant

File: eq-author-api/src/validation/schemas/entities.json (hunk 2 above)
Comment (pricea1) [reply eq-author-api/src/validation/schemas/entities.json]: Now restricted to TextField input type. We'll need to remove this restriction once we want it to apply to all label fields

File: eq-author/src/App/page/Design/answers/MultipleChoiceAnswer/__snapshots__/MultipleChoiceAnswer.test.js.snap (hunk 4 above)
Comment (pricea1) [reply eq-author/src/App/page/Design/answers/MultipleChoiceAnswer/__snapshots__/MultipleChoiceAnswer.test.js.snap]: yup, I hadn't defined the displayName of withValidationError HoC. Now fixed and snapshots updated

**Comment by pricea1** [comment]
[Validator - Basic Answer (4)](https://trello.com/c/pu89dqzO/1102-validator-basic-answer-4)

File: eq-author/src/components/Forms/WrappingInput/index.js (hunk 3 above)
Comment (hamishtaplin) [reply eq-author/src/components/Forms/WrappingInput/index.js]: The issue is styled components is passing this onto the DOM. We should solve that problem instead by wrapping the `TextArea` in another component that only passes the valid props on:

https://github.com/ONSdigital/eq-author-app/compare/prototype-survey-intro?expand=1#diff-02621415b811b4593f1d08f24d9b9c81R16

**Comment by hamishtaplin** [comment]
Not sure if this is an issue introduced here but when there is a validation error, focus is moved to the `body`. You can test this by tabbing out of the `label` of any answer that has a description.

**Comment by hamishtaplin** [comment]
We should consider adding [Aria attributes](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute) to invalid elements.

File: eq-author/src/enhancers/withValidationError.js [hunk 5]
```diff
@@ -14,14 +14,16 @@ const withValidationError = entityPropName => WrappedComponent => {
...
+
     getValidationError = ({ field, label }) => {
       const {
         [entityPropName]: { validationErrorInfo, isNew },
       } = this.props;
 
       const messages = validationErrorInfo && validationErrorInfo.errors;
 
-      if (!messages || isNew) {
+      if (!messages || messages.length === 0 || isNew) {
```
Comment (hamishtaplin) [inline eq-author/src/enhancers/withValidationError.js]: `isEmpty(messages) || isNew`?

File: eq-author/src/apollo/resolvers/index.test.js [hunk 6]
```diff
@@ -16,71 +17,114 @@ describe("resolvers", () => {
...
+      resolvers.Mutation.updateQuestionPage(updateQuestionnaire, {}, { cache });
 
-    expect(cacheData.newPagesList).toMatchObject(newPages);
+      expect(cacheData.newEntityList).toMatchObject(newUpdatePages);
+    });
 
-    const updateQuestionnaire = {
-      updateQuestionPage: { id: 5 },
-    };
+    it("should initalise pages when creating a new section", () => {
```
Comment (heofs) [inline eq-author/src/apollo/resolvers/index.test.js]: *initialise

File: eq-author/src/components/Forms/WrappingInput/index.js (hunk 3 above)
Comment (pricea1) [reply eq-author/src/components/Forms/WrappingInput/index.js]: Changed to use wrapper and boolean prop as suggested

**Comment by pricea1** [comment]
> Not sure if this is an issue introduced here but when there is a validation error, focus is moved to the `body`. You can test this by tabbing out of the `label` of any answer that has a description.

Not introduced in this PR but the one that added validation to question title. Now fixed.

File: eq-author/src/apollo/resolvers/index.test.js (hunk 6 above)
Comment (pricea1) [reply eq-author/src/apollo/resolvers/index.test.js]: Fixed

File: eq-author/src/enhancers/withValidationError.js (hunk 5 above)
Comment (pricea1) [reply eq-author/src/enhancers/withValidationError.js]: Thats better. I've updated it.

**Comment by pricea1** [comment]
> We should consider adding [Aria attributes](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute) to invalid elements.

aria-invalid added to answer label.  draft-js-plugins-editor Rich text editor used in question title doesn't appear to support aria-invalid: [Allowed props for draft-js-plugins-editor](https://github.com/facebook/draft-js/blob/master/src/component/base/DraftEditorProps.js)

File: eq-author-api/src/validation/schemas/entities.json [hunk 7]
```diff
@@ -1,31 +1,52 @@
...
+        },
+        "pages": {
+          "type": "array",
+          "items": {
+            "$ref": "entities.json#/definitions/page"
+          }
         }
       },
-      "required": ["id", "title"]
+      "required": ["id", "title", "pages"]
```
Comment (tgandrews) [inline eq-author-api/src/validation/schemas/entities.json]: I don't think `title` on section is required

File: eq-author-api/src/validation/schemas/entities.json [hunk 8]
```diff
@@ -1,31 +1,52 @@
...
+      },
+      "if": { "properties": { "type": { "enum": ["TextField", "Number"] } } },
+      "then": {
+        "properties": {
+          "label": {
+            "$ref": "definitions.json#/definitions/populatedString"
+          }
+        }
+      },
+      "required": ["id", "label"]
```
Comment (tgandrews) [inline eq-author-api/src/validation/schemas/entities.json]: Not all answers require labels

File: eq-author/src/apollo/resolvers/index.js [hunk 9]
```diff
@@ -1,73 +1,88 @@
...
+const initialiseMultiplePages = (pages, cache) => {
+  pages.forEach(questionPage => {
+    addNewEntity(PAGES, questionPage.id, cache);
+  });
 };
 
-const isNewQuestionPage = (questionPage, args, { cache }) => {
-  const { newPagesList } = cache.readQuery({
-    query: GET_NEW_PAGES_LIST,
+const isNewEntity = type => (questionPage, args, { cache }) => {
```
Comment (tgandrews) [inline eq-author/src/apollo/resolvers/index.js]: Is it always a page here? If not it shouldn't be `questionPage`

File: eq-author/src/components/withEntityEditor/index.js [hunk 10]
```diff
@@ -64,7 +64,7 @@ const withEntityEditor = entityPropName => WrappedComponent => {
     };
 
     handleUpdate = () => {
-      if (!this.dirtyField) {
+      if (!this.dirtyField && !this.entity.isNew) {
```
Comment (tgandrews) [inline eq-author/src/components/withEntityEditor/index.js]: Can we add a unit test around this

File: eq-author-api/src/validation/schemas/entities.json [hunk 11]
```diff
@@ -1,31 +1,52 @@
...
       },
-      "required": ["id", "title", "pages"]
+      "required": ["id", "title"]
+    },
+    "answer": {
+      "type": "object",
+      "properties": {
+        "id": { "type": "string" }
+      },
+      "if": { "properties": { "type": { "enum": ["TextField", "Number"] } } },
```
Comment (tgandrews) [inline eq-author-api/src/validation/schemas/entities.json]: There are Currency, Percentage and TextArea as well

Comment (pricea1) [reply eq-author-api/src/validation/schemas/entities.json]: The trello ticket specified Numeric & Text fields only:
[https://trello.com/c/pu89dqzO/1102-validator-basic-answer-4](https://trello.com/c/pu89dqzO/1102-validator-basic-answer-4)

Comment (tgandrews) [reply eq-author-api/src/validation/schemas/entities.json]: Currency and Percentage are numeric fields. And we have two text fields - text field and text area.

Comment (pricea1) [reply eq-author-api/src/validation/schemas/entities.json]: Now changed to include TextField, Curreny & Percentage

File: eq-author/src/components/withEntityEditor/index.js (hunk 10 above)
Comment (pricea1) [reply eq-author/src/components/withEntityEditor/index.js]: Done

File: eq-author/src/apollo/resolvers/index.js (hunk 9 above)
Comment (pricea1) [reply eq-author/src/apollo/resolvers/index.js]: No, it was before this PR but I'd missed renaming the variable. Done now.

**Pull request merged**